django_sharding_library/utils.py

```diff
@@ -0,0 +1,32 @@
+from django.db import connections, DatabaseError, transaction
+from django.conf import settings
+from django_sharding_library.sql import postgres_shard_id_function_sql
+from django.db.models import signals
+
+
+def create_postgres_global_sequence(sequence_name, db_alias, reset_sequence=False):
+    cursor = connections[db_alias].cursor()
+    sid = transaction.savepoint(db_alias)
+    try:
+        cursor.execute("CREATE SEQUENCE %s;" % sequence_name)
+    except DatabaseError:
+        transaction.savepoint_rollback(sid, using=db_alias)
+        if reset_sequence:
+            cursor.execute("SELECT setval('%s', 1, false)" % (sequence_name,))
+    else:
+        transaction.savepoint_commit(sid, using=db_alias)
+    cursor.close()
+
+
+def create_postgres_shard_id_function(sequence_name, db_alias, shard_id):
+    cursor = connections[db_alias].cursor()
+    cursor.execute(postgres_shard_id_function_sql % {'shard_epoch': settings.SHARD_EPOCH,
+                                                     'shard_id': shard_id,
+                                                     'sequence_name': sequence_name})
+    cursor.close()
+
+
+def register_migration_signal_for_model_receiver(model, function, dispatch_uid=None):
+    signals.pre_migrate.connect(function, sender=model, dispatch_uid=dispatch_uid)
```

Comment:
It will get a DatabaseError if it fails. Its in pre-migrate because its the only common area to add a migration step without editing the migration file itself after its produced, which I don't think is a good feature for a library since it adds complexity to using the library that can be taken care of automatically in the pre migration.

Also, it has to be in the pre migrate because if one of the postgres fields gets created and the stored procedure doesnt exist yet, it will raise a programmingerror. Doing it in a post migration or anywhere else doesn't work.